src/emu.h

```diff
@@ -0,0 +1,111 @@
+#include <iostream>
+
+using byte_t = uint8_t;
+using word_t = uint16_t;
+template<typename T> bool IsPtrAligned(word_t ptr);
+template<> bool IsPtrAligned<byte_t>(word_t ptr) { return true; }
+//template<> bool IsPtrAlgined<word_t>(word_t ptr) { return !(ptr % sizeof(word_t)); }
+
+struct Emu {
+	enum struct TrapVec : byte_t {
+		USER_DEV	= 0000,
+		ODD_ADDR	= 0004,
+		ILL_INSTR	= 0004,
+		STK_RED		= 0004,
+		RES_INSTR	= 0010,
+		TRACE		= 0014,
+		POWER_FAIL	= 0024,
+		EMT		= 0030,
+		TRAP		= 0034,
+		MM_TRAP		= 0250,
+	} trapVec;
+	bool trapPending = false;
+	void RaiseTrap(TrapVec const trap) { trapVec = trap; trapPending = true; };
+
+	enum struct GenRegId : byte_t {
+		R0	= 00,
+		R1	= 01,
+		R2	= 02,
+		R3	= 03,
+		R4	= 04,
+		R5	= 05,
+		R6	= 06,
+		R7	= 07,
+		PC	= R7,
+		HW_SP	= R6,
+		MAX	= 07,
+	};
+	word_t genRegfile[static_cast<byte_t>(GenRegId::MAX)];
+
+	struct Memory {
+		Emu &emu;
+		byte_t *dummy;
+		word_t const dummy_start = 00377;
+		word_t const dummy_end   = 01377;
+		Memory(Emu &_emu) : emu(_emu) { dummy = new byte_t[dummy_end - dummy_start + 1]; }
+		~Memory() { delete[] dummy; }
+		bool IsMapped(word_t ptr) { return ptr >= dummy_start && ptr <= dummy_end; }
+	private:
+		template<typename T> bool CheckPtr(word_t ptr) {
+			if (!IsMapped(ptr)) {
+				emu.RaiseTrap(TrapVec::MM_TRAP);
+				return false;
+			}
+			if (!IsPtrAligned<T>(ptr)) {
+				emu.RaiseTrap(TrapVec::ODD_ADDR);
+				return false;
+			}
+			return true;
+		}
+	public:
+		template<typename T> bool Load(word_t ptr, T &val) {
+			if (!CheckPtr<T>(ptr))
+				return false;
+			val = *(reinterpret_cast<T*>(&dummy[ptr-dummy_start]));
+			return true;
+		}
+	} memory;
+
+	struct InstrBase {
+		virtual void Execute(struct Emu &emu)=0;
+		virtual std::ostream &operator<<(std::ostream &os)=0;
+	};
+
+	struct UnknownInstr : public InstrBase {
+		UnknownInstr() {}
+		void Execute(struct Emu &emu) { emu.RaiseTrap(TrapVec::ILL_INSTR); }
+		std::ostream &operator<<(std::ostream &os) { return os << "Unknown instr"; }
+	};
+
+	union __attribute__((may_alias)) Instr {
```

Comment:
I've peeked at the gcc docs regarding `may_alias` and understood very little. Throw a couple of words explaining the purpose of this attribute here please.

Reply:
reinterpret_cast in getBase() method may fail if GCC will not produce union Instr suitable for aliasing. So I've placed may_alias attribute.